### ISSUE TITLE
Start using a dependency and/or build cache in CI.

### ISSUE THREAD

**Issue by clarketm** [istio/test-infra#2425]
Start using a shared dependency and/or build cache for `bazel` and `go` to improve build and test efficiency in CI.

**Comment by howardjohn** [comment]
Testing go.mod and build caching:

go mod download
* No cache: 1min
* Cold cache: 2m48s (slower because the disk is slower)
* warm cache: 2s


For build
* no cache: 54s
* cold cache: 58s
* warm cache: 20s

cost: base is $200/month for 1TB at standard speeds. We can also do $750 for premium which is 2.5TB and quite a bit faster

One build uses 500mb in go build cache, 1.3gb module caching. 

> The go command periodically deletes cached data that has not been used recently.

I don't think there is automatic module cache cleanup.

alternative for go modules is we can pre-warm the module cache inside the docker image itself, at the cost of +1.3gb to the image


This is all done with Google Filestore which is a managed NFS server. Test yamls here:

```yaml

apiVersion: apps/v1
kind: Deployment
metadata:
  name: shell
spec:
  selector:
    matchLabels:
      app: shell
  template:
    metadata:
      labels:
        app: shell
    spec:
      containers:
      - name: shell
        imagePullPolicy: Always
        image: gcr.io/istio-testing/build-tools:master-2020-02-14T13-09-14
        args:
          - /bin/sleep
          - infinity
        volumeMounts:
        - mountPath: /go/pkg
          subPath: gomod
          name: mypvc
        - mountPath: /gocache
          subPath: buildcache
          name: mypvc
      volumes:
      - name: mypvc
        persistentVolumeClaim:
          claimName: fileserver-claim
          readOnly: false
---
apiVersion: v1
kind: PersistentVolume
metadata:
  name: fileserver
spec:
  capacity:
    storage: 1T
  accessModes:
  - ReadWriteMany
  nfs:
    path: /gomodtest1
    server: IP-OF-FILESTORE
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: fileserver-claim
spec:
  accessModes:
  - ReadWriteMany
  storageClassName: ""
  volumeName: fileserver
  resources:
    requests:
      storage: 5Gi
```


Basically we can have one filestore, and then have subdirectories for each cache we want (go mod, go build, bazel, ...)

**Comment by howardjohn** [comment]
caching docker images is abysmally slow since overlayv2 is not supported over NFS, and I think it will break if two daemons use it at once, so that option is a no-go. If we want to cache images we should use a local registry, or store a bunch of tarbals in the NFS drive and `docker load` them

**Comment by howardjohn** [comment]
seems reasonable to do the following:
- set up the filestore in istio-testing
- add PV and PVC to cluster
- create a job that is hidden/optional in istio that runs `make build` with the volume mounted and we track the speed and stability of the job for a bit to see what happens

Any objections? I can work on this if not? The main thing I'm not sure of is if there is a more standard way to do this in prow that we should be relying on rather than DIY.

The reason to do it with the filestore is as far as I can tell it's the only ReadWriteMany type we have access to

**Comment by clarketm** [comment]
> Any objections? I can work on this if not? The main thing I'm not sure of is if there is a more standard way to do this in prow that we should be relying on rather than DIY.

sgtm. No, I do not have any objections. The only remote caching service that I am aware of is [greenhouse](https://github.com/kubernetes/test-infra/tree/master/greenhouse), but I have not used it personally so it may not be suitable for the need.

**Comment by howardjohn** [comment]
Ok I think that is highly bazel specific, ill send out a PR for this approach soon.

**Comment by howardjohn** [comment]
This is in use now
